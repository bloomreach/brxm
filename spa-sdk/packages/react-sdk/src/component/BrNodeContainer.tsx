/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import {
  Container,
  TYPE_CONTAINER_INLINE,
  TYPE_CONTAINER_NO_MARKUP,
  TYPE_CONTAINER_ORDERED_LIST,
  TYPE_CONTAINER_UNORDERED_LIST,
} from '@bloomreach/spa-sdk';
import { BrNodeComponent } from './BrNodeComponent';
import { BrContainerBox, BrContainerInline, BrContainerNoMarkup, BrContainerOrderedList, BrContainerUnorderedList } from '../cms';

export class BrNodeContainer extends BrNodeComponent<Container> {
  protected getMapping() {
    return this.props.component.getType();
  }

  protected fallback() {
    switch (this.props.component.getType()) {
      case TYPE_CONTAINER_INLINE: return <BrContainerInline {...this.props} />;
      case TYPE_CONTAINER_NO_MARKUP: return <BrContainerNoMarkup {...this.props} />;
      case TYPE_CONTAINER_ORDERED_LIST: return <BrContainerOrderedList {...this.props} />;
      case TYPE_CONTAINER_UNORDERED_LIST: return <BrContainerUnorderedList {...this.props} />;
      default: return <BrContainerBox {...this.props} />;
    }
  }
}
