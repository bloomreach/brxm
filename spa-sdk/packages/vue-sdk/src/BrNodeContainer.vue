<!--
  Copyright 2020 Hippo B.V. (http://www.onehippo.com)

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
  -->

<template>
  <component
    v-if="component.getType() in mapping"
    :is="mapping[component.getType()]"
    :component="component"
    :page="page"
  >
    <slot />
  </component>

  <br-container-inline v-else-if="component.getType() === TYPE_CONTAINER_INLINE" :component="component" :page="page">
    <slot />
  </br-container-inline>

  <br-container-no-markup
    v-else-if="component.getType() === TYPE_CONTAINER_NO_MARKUP"
    :component="component"
    :page="page"
  >
    <slot />
  </br-container-no-markup>

  <br-container-ordered-list
    v-else-if="component.getType() === TYPE_CONTAINER_ORDERED_LIST"
    :component="component"
    :page="page"
  >
    <slot />
  </br-container-ordered-list>

  <br-container-unordered-list
    v-else-if="component.getType() === TYPE_CONTAINER_UNORDERED_LIST"
    :component="component"
    :page="page"
  >
    <slot />
  </br-container-unordered-list>

  <br-container-box v-else :component="component" :page="page">
    <slot />
  </br-container-box>
</template>

<script lang="ts">
import {
  Container,
  Page,
  TYPE_CONTAINER_INLINE,
  TYPE_CONTAINER_NO_MARKUP,
  TYPE_CONTAINER_ORDERED_LIST,
  TYPE_CONTAINER_UNORDERED_LIST,
} from '@bloomreach/spa-sdk';
import { Component, Inject, Vue } from 'vue-property-decorator';
import BrContainerBox from './BrContainerBox.vue';
import BrContainerInline from './BrContainerInline.vue';
import BrContainerNoMarkup from './BrContainerNoMarkup.vue';
import BrContainerOrderedList from './BrContainerOrderedList.vue';
import BrContainerUnorderedList from './BrContainerUnorderedList.vue';

@Component({
  components: {
    BrContainerBox,
    BrContainerInline,
    BrContainerNoMarkup,
    BrContainerOrderedList,
    BrContainerUnorderedList,
  },
  computed: {
    component(this: BrNodeContainer) {
      return this.component$();
    },
    mapping(this: BrNodeContainer) {
      return this.mapping$();
    },
    page(this: BrNodeContainer) {
      return this.page$();
    },
  },
  data: () => ({
    TYPE_CONTAINER_INLINE,
    TYPE_CONTAINER_NO_MARKUP,
    TYPE_CONTAINER_ORDERED_LIST,
    TYPE_CONTAINER_UNORDERED_LIST,
  }),
})
export default class BrNodeContainer extends Vue {
  @Inject() private component$!: () => Container;

  @Inject() private mapping$!: () => Record<string, Vue.Component>;

  @Inject() private page$!: () => Page;
}
</script>
