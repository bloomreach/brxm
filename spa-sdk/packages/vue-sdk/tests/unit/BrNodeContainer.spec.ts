/*
 * Copyright 2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Wrapper, shallowMount } from '@vue/test-utils';
import { Component, Prop, Provide, Vue } from 'vue-property-decorator';
import {
  Container,
  Page,
  TYPE_CONTAINER_BOX,
  TYPE_CONTAINER_INLINE,
  TYPE_CONTAINER_NO_MARKUP,
  TYPE_CONTAINER_ORDERED_LIST,
  TYPE_CONTAINER_UNORDERED_LIST,
} from '@bloomreach/spa-sdk';
import BrNodeContainer from '@/BrNodeContainer.vue';

@Component({ template: '<div />' })
class BrPage extends Vue {
  @Prop() component!: Container;

  @Provide() component$() {
    return this.component;
  }

  @Prop() mapping!: Record<string, Vue.Component>;

  @Provide() mapping$() {
    return this.mapping;
  }

  @Prop() page!: Page;

  @Provide() page$() {
    return this.page;
  }
}

@Component
class SomeContainer extends Vue {
  @Prop() component!: Container;

  @Prop() page!: Page;
}

describe('BrNodeContainer', () => {
  let component: jest.Mocked<Container>;
  let mapping: Record<string, Vue.Component>;
  let page: jest.Mocked<Page>;
  let provide: () => unknown;
  let parent: Wrapper<Vue>;

  beforeEach(() => {
    component = ({
      getType: jest.fn(),
    } as unknown) as typeof component;
    mapping = {};
    page = ({} as unknown) as typeof page;

    parent = shallowMount(BrPage, { propsData: { component, mapping, page } });
    provide = (parent.vm.$options.provide as typeof provide).bind(parent.vm);
  });

  afterEach(() => {
    parent.destroy();
  });

  describe('render', () => {
    it('should render a mapped container', async () => {
      component.getType.mockReturnValue('custom' as ReturnType<typeof component.getType>);
      mapping.custom = SomeContainer;

      const wrapper = shallowMount(BrNodeContainer, { provide, slots: { default: ['<div />'] } });
      await wrapper.vm.$nextTick();

      const props = wrapper.findComponent(SomeContainer).props();

      expect(wrapper.html()).toMatchSnapshot();
      expect(props.component).toBe(component);
      expect(props.page).toEqual(page);
    });

    it.each`
      type
      ${TYPE_CONTAINER_BOX}
      ${TYPE_CONTAINER_INLINE}
      ${TYPE_CONTAINER_NO_MARKUP}
      ${TYPE_CONTAINER_ORDERED_LIST}
      ${TYPE_CONTAINER_UNORDERED_LIST}
    `('should render a container based on the type', async ({ type }) => {
      component.getType.mockReturnValue(type);

      const wrapper = shallowMount(BrNodeContainer, { provide, slots: { default: ['<div />'] } });
      await wrapper.vm.$nextTick();

      expect(wrapper.html()).toMatchSnapshot();
    });
  });
});
