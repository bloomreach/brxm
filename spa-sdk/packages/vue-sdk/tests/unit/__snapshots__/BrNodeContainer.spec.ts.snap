// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrNodeContainer render should render a container based on the type 1`] = `
<br-container-box-stub page="[object Object]" component="[object Object]">
  <div></div>
</br-container-box-stub>
`;

exports[`BrNodeContainer render should render a container based on the type 2`] = `
<br-container-inline-stub page="[object Object]" component="[object Object]">
  <div></div>
</br-container-inline-stub>
`;

exports[`BrNodeContainer render should render a container based on the type 3`] = `
<br-container-no-markup-stub component="[object Object]" page="[object Object]">
  <div></div>
</br-container-no-markup-stub>
`;

exports[`BrNodeContainer render should render a container based on the type 4`] = `
<br-container-ordered-list-stub page="[object Object]" component="[object Object]">
  <div></div>
</br-container-ordered-list-stub>
`;

exports[`BrNodeContainer render should render a container based on the type 5`] = `
<br-container-unordered-list-stub page="[object Object]" component="[object Object]">
  <div></div>
</br-container-unordered-list-stub>
`;

exports[`BrNodeContainer render should render a mapped container 1`] = `
<somecontainer-stub component="[object Object]" page="[object Object]">
  <div></div>
</somecontainer-stub>
`;
